bndtools.core/src/bndtools/central/Central.java

```diff
@@ -204,39 +203,52 @@ public static Workspace getWorkspace() throws Exception {
             if (ws != null) { // early check for workspace
                 return ws;
             }
-            try {
-                Workspace.setDriver(Constants.BNDDRIVER_ECLIPSE);
-                Workspace.addGestalt(Constants.GESTALT_INTERACTIVE, new Attrs());
+            ws = initializeWorkspace();
+            // The workspace has been initialized fully, set the field now
+            workspace = ws;
+        }
+        workspaceQueue.resolve(ws); // notify onWorkspaceInit callbacks
```

Comment:
I had a look at the code again. As far as I understand the Promise, workspaceQueue is resolved with the new workspace instance, not the old one. This causes the callbacks registered in onWorkspaceInit to be fired with the same new instance.

Also workspaceQueue does not need to change over time, since it only holds Functions that have no relation to the workspace. Again, the workspace is passed during resolve.

The only case which was not handled for the callbacks is a null workspace. I've added a check that the callbacks will only be called if workspace is set.
The current (master) implementation cannot handle null workspaces...it will just end up in an exception which gets logged.